### PR TITLE
elastic queue use case test based on ARTEMIS-3365 and ARTEMIS-3569

### PR DESCRIPTION
**Pull request by gtully**
scenario - avoid paging, if address is full chain another broker and produce to the head, consume from the tail using producer and consumer roles to partition connections. When tail is drained, drop it.
 - adds a option to treat an idle consumer as slow
 - adds basic support for credit based address blocking ARTEMIS-2097
 - adds some more visiblity to address memory usage and balancer attribute modifier operations

### CODE REVIEW

**Comment by gtully** [comment]
the block feature needs some work, I don't fully have a handle on how flows are collected. I need some help here. The intent is to have an address where consumers can drain and producers are blocked rather than whacked. The use case is partitioning an address by producers/consumers, but supporting the scenario where a single connection has both producers and consumers!
I would like to block a producer via 0 link credit for that address, but the flow control is based on refreshing credit atm. I may be in the wrong place altogether :-)

**Comment by gtully** [comment]
on more investigation, in the absence of doing a drain, we respect existing credit and track requests up to the credit limits. This works for the already limited case needing to get an exception (FAIL strategy) and allows resumption/unblocking to renew credit when there are no other constraints. Eventually such a producer will run out of credit and block.
For new links/senders - there is no new credit grant which forces blocking on credit and this is perfect. Senders can stay blocked or use a send timeout to error out of that state.
In short, I think the blocking is in the right place.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 1]
```diff
@@ -1153,6 +1161,31 @@ public boolean isFull() {
       return maxSize > 0 && getAddressSize() > maxSize || pagingManager.isGlobalFull();
    }
 
+
+   @Override
+   public int getAddressLimitPercent() {
+      final long currentUsage = getAddressSize();
+      if (currentUsage != 0) {
+         if (maxSize > 0) {
+            return Math.toIntExact((currentUsage / maxSize) * 100);
```
Comment (brusdev) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:1170]: Dividing currentUsage by maxSize should always return 0

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 2]
```diff
@@ -1153,6 +1161,31 @@ public boolean isFull() {
...
+         } else if (pagingManager.isUsingGlobalSize()) {
+            return Math.toIntExact((currentUsage / pagingManager.getGlobalSize()) * 100);
```
Comment (brusdev) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:1172]: Same as previous comment dividing currentUsage by pagingManager.getGlobalSize() should always return 0

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/client/SlowConsumerTest.java [hunk 3]
```diff
@@ -612,6 +612,45 @@ public void testMinuteSurviving() throws Exception {
...
+         producer.send(session.createMessage(true));
+      }
+      session.commit();
+
+      ConcurrentHashSet<ClientMessage> receivedMessages = new ConcurrentHashSet<>();
+      FixedRateConsumer consumer = new FixedRateConsumer(10, MESSAGES_PER_SECOND, receivedMessages, sf, QUEUE, 0);
+      consumer.start();
+
+      Queue queue = server.locateQueue(QUEUE);
+      Wait.assertEquals(1, queue::getConsumerCount);
```
Comment (brusdev) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/client/SlowConsumerTest.java:646]: This could be time sensitive, what about replacing FixedRateConsumer with a manual consumer and check consumerCount before receiving the message?

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/BrokerBalancerControlTest.java [hunk 4]
```diff
@@ -141,6 +141,27 @@ public void testGetLocalTarget() throws Exception {
 
       CompositeData connectorData = (CompositeData)targetData.get("connector");
       Assert.assertNull(connectorData);
+
+      assertNull(brokerBalancerControl.getLocalTargetFilter());
```
Comment (brusdev) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/BrokerBalancerControlTest.java:145]: why not creating other tests for those new BrokerBalancerControl methods?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 5]
```diff
@@ -695,6 +697,14 @@ public boolean checkMemory(final Runnable runWhenAvailable) {
    @Override
    public boolean checkMemory(boolean runOnFailure, final Runnable runWhenAvailable) {
 
+      if (blocked) {
+         if (runWhenAvailable != null) {
+            onMemoryFreedRunnables.add(AtomicRunnable.checkAtomic(runWhenAvailable));
```
[the remark this answers is not in the document]
Comment (brusdev) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:702]: maybe checkReleaseMemory could check if it is `blocked`

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java [hunk 6]
```diff
@@ -0,0 +1,697 @@
...
+      assertTrue(Wait.waitFor(() -> eqProducer.connectionCount.get() > 1));
+
+      // operator mode, poll queue control - to allow producer to continue, activate next broker in the 'chain'
+      AddressControl addressControl0 = (AddressControl) ManagementControlHelper.createProxy(node0NameBuilder.getAddressObjectName(qNameSimple), AddressControl.class, mBeanServer);
+
+      AddressControl finalAddressControl = addressControl0;
+      assertTrue(Wait.waitFor(() -> {
+         int usage = finalAddressControl.getAddressLimitPercent();
+         System.out.println("Node0 usage % " + usage);
+         return usage == 100;
```
Comment (brusdev) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java:424]: I guess usage percent could never be 100

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java [hunk 7]
```diff
@@ -0,0 +1,697 @@
...
+      eqConsumer.start();
+
+      // verify consumer reconnects on no messages
+      assertTrue(Wait.waitFor(() -> eqConsumer.connectionCount.get() > 1));
+
+      EQProducer eqProducer = new EQProducer(urlForNodes(nodes));
+      eqProducer.start();
+
+      // verify producer reconnects on fail full!
+      assertTrue(Wait.waitFor(() -> eqProducer.connectionCount.get() > 1));
```
Comment (brusdev) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java:415]: why not using Wait.assertTrue ?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gemmellr) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:702]: unBlock() would need to do the reverse too I guess, check if `blocking`.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java (hunk 1 above)
Comment (gemmellr) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:1170]: Or 1, if currentUsage is the same (or bigger but < double).

Phrased another way: unit tests missing with 0% !=  x != 100% checks also?

(the currentUsage != 0 could probably go away too)

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java [hunk 8]
```diff
@@ -0,0 +1,697 @@
...
+import org.junit.Test;
+
+public class ElasticQueueTest extends ActiveMQTestBase {
+
+   static final String qName = "EQ";
+   static final SimpleString qNameSimple = SimpleString.toSimpleString(qName);
+
+   final int base_port = 61616;
+   final int opTimeoutMillis = 2000;
+   final Stack<EmbeddedActiveMQ> nodes = new Stack<>();
```
Comment (gemmellr) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java:74]: Does anything clean these up when the test fails mid-way?

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java [hunk 9]
```diff
@@ -0,0 +1,697 @@
...
+      public void onConsumerClosed(MessageConsumer messageConsumer, Throwable throwable) {
+      }
+
+      @Override
+      public void onProducerClosed(MessageProducer messageProducer, Throwable throwable) {
+      }
+   }
+
+   // slow consumer
+   class EQConsumer extends Thread {
```
Comment (gemmellr) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java:135]: Probably nicer to use an executor which can then be cleaned up afterwards

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java (hunk 7 above)
Comment (gemmellr) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java:415]: Also the test has an overall 20sec timeout, but this single check (and all the other ones) has a 30sec timeout. Seems like the checks should be gated by how long they are reasonably expected to take.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:702]: good catch, the do need to coordinate.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 10]
```diff
@@ -1153,6 +1161,31 @@ public boolean isFull() {
...
+      return 0;
+   }
+
+   @Override
+   public void block() {
+      blocked = true;
+   }
+
+   @Override
+   public void unBlock() {
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:1184]: thanks!

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java [hunk 11]
```diff
@@ -85,6 +89,34 @@ public void testCreditsAreAllocatedOnceOnLinkCreated() throws Exception {
       }
    }
 
+   @Test(timeout = 60000)
+   public void testCreditsAreNotAllocatedOnceOnLinkCreatedWhileBlockedAndWhenUnBlocked() throws Exception {
```
[the remark this answers is not in the document]
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java:93]: better

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java [hunk 12]
```diff
@@ -130,6 +162,51 @@ public void testAddressIsBlockedForOtherProdudersWhenFull() throws Exception {
       assertTrue(addressSize >= MAX_SIZE_BYTES_REJECT_THRESHOLD);
    }
 
+   @Test(timeout = 10000)
+   public void testSendHangsWhenBlockedAndNotAfterUnBlocked() throws Exception {
```
[the remark this answers is not in the document]
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java:166]: better

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java [hunk 13]
```diff
@@ -130,6 +162,51 @@ public void testAddressIsBlockedForOtherProdudersWhenFull() throws Exception {
...
+            } catch (JMSException ignored) {
+            } finally {
+               done.countDown();
+            }
+         }
+      }).start();
+
+      assertTrue(running.await(5, TimeUnit.SECONDS));
+
+      assertFalse(done.await(1, TimeUnit.SECONDS));
```
[the remark this answers is not in the document]
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java:198]: true

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java [hunk 14]
```diff
@@ -85,6 +89,34 @@ public void testCreditsAreAllocatedOnceOnLinkCreated() throws Exception {
...
+   public void testCreditsAreNotAllocatedOnceOnLinkCreatedWhileBlockedAndWhenUnBlocked() throws Exception {
+      AmqpClient client = createAmqpClient(new URI(singleCreditAcceptorURI));
+      AmqpConnection connection = addConnection(client.connect());
+
+      try {
+         AddressControl addressControl = ManagementControlHelper.createAddressControl(SimpleString.toSimpleString(getQueueName()), mBeanServer);
+         addressControl.block();
+         AmqpSession session = connection.createSession();
+         final AmqpSender sender = session.createSender(getQueueName());
+         assertEquals("Should get 0 credit", 0, sender.getSender().getCredit());
```
[the remark this answers is not in the document]
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java:102]: thanks for that pointer

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/balancing/BrokerBalancer.java [hunk 15]
```diff
@@ -201,4 +201,16 @@ public TargetResult getTarget(String key) {
 
       return result != null ? result : TargetResult.REFUSED_UNAVAILABLE_RESULT;
    }
+
+   public void setLocalTargetFilter(String regExp) {
+      if (regExp == null || regExp.isBlank()) {
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/balancing/BrokerBalancer.java:206]: no need for that, well spotted!

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/BrokerBalancerControlTest.java (hunk 4 above)
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/BrokerBalancerControlTest.java:145]: would be better, thanks!

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/balancing/BrokerBalancer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (gemmellr) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/balancing/BrokerBalancer.java:206]: I wasnt suggesting not to use it, in case thats what it seemed like.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java (hunk 1 above)
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java:1170]: that does need work and tests, on it, thanks!

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java [hunk 16]
```diff
@@ -85,6 +89,34 @@ public void testCreditsAreAllocatedOnceOnLinkCreated() throws Exception {
...
+
+         addressControl.unBlock();
+         assertTrue("Should now get issued one credit", Wait.waitFor(new Wait.Condition() {
+            @Override
+            public boolean isSatisfied() throws Exception {
+               return 1 == sender.getSender().getCredit();
+            }
+         }));
```
[the remark this answers is not in the document]
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpFlowControlTest.java]: fair, it will speed up the test a little

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java (hunk 6 above)
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java:431]: good catch, the test did indeed need some work with the fixed getAddressLimitPercent :-)

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java (hunk 9 above)
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java]: agree.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java (hunk 8 above)
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java:76]: sorted

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 17]
```diff
@@ -126,6 +126,8 @@
 
    private volatile boolean blocking = false;
 
+   private volatile boolean blocked = false;
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java]: done

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java [hunk 18]
```diff
@@ -695,6 +697,14 @@ public boolean checkMemory(final Runnable runWhenAvailable) {
    @Override
    public boolean checkMemory(boolean runOnFailure, final Runnable runWhenAvailable) {
 
+      if (blocked) {
+         if (runWhenAvailable != null) {
+            onMemoryFreedRunnables.add(AtomicRunnable.checkAtomic(runWhenAvailable));
+         }
+         ActiveMQServerLogger.LOGGER.blockingMessageProduction(address, sizeInBytes.get(), maxSize, pagingManager.getGlobalSize());
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java]: agree. I have added its own log on state change, the audit logger on operations also covers it but I think it is good to have an indication in the regular server log for something with large repercussions like blocking

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java (hunk 2 above)
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java]: thanks, sorted.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/client/SlowConsumerTest.java (hunk 3 above)
Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/client/SlowConsumerTest.java:646]: is it the locate queue that may not yet be present?

Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/client/SlowConsumerTest.java:646]: I suppose there is no need for a fixed rate consumer tho, it is irrelevant to the test. a simple consumer would suffice

Comment (gtully) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/client/SlowConsumerTest.java]: I have tidied it up a bit, not 100% I captured your feedback!

**Comment by gtully** [comment]
draft till I rebase and possibly pull in a pid controller

**Comment by gtully** [comment]
going to leave the control loop for another day and have a simple isFull check. there is lots to study in https://en.wikipedia.org/wiki/PID_controller#Response_to_disturbances

**Comment by gtully** [comment]
any more feedback?

**Comment by gemmellr** [comment]
Not from me.
